packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -459,12 +352,261 @@ export class OdspDocumentStorageService implements IDocumentStorageService {
+            const blobsIdToPathMap: Map<string, string> = new Map();
+            let appCommit: string | undefined;
+            let appTree: string | undefined;
+
+            for (const [key, treeVal] of this.treesCache.entries()) {
+                if (!appCommit && !appTree) {
+                    for (const entry of treeVal.entries) {
+                        if (entry.path === ".app") {
+                            if (entry.type === "commit") {
+                                // This is the unacked handle of the latest summary generated.
+                                appCommit = entry.id;
+                            }
+                            if (entry.type === "tree") {
+                                appTree = entry.id;
+                            }
+                            break;
+                        }
+                    }
+                    assert(appCommit || appTree); // .app commit or tree should be first entry in first entry.
+                }
+                for (const entry of treeVal.entries) {
+                    if (entry.type === "blob") {
+                        blobsIdToPathMap.set(entry.id, key === appCommit ? `/.app/${entry.path}` : `/${entry.path}`);
+                    }
+                }
+            }
+
+            // Populate the cache with paths from id-to-path mapping.
+            for (const blob of this.blobCache.values()) {
+                const path = blobsIdToPathMap.get(blob.id);
+                // If this is the first container that was created for the service, it cannot be
+                // the summarizing container (because the summarizing container is always created
+                // after the main container). In this case, we do not need to do any hashing
+                if (path) {
+                    // Schedule the hashes for later, but keep track of the tasks
+                    // to ensure they finish before they might be used
+                    const hashP = hashFile(IsoBuffer.from(blob.content, blob.encoding)).then((hash: string) => {
+                        this.blobsShaToPathCache.set(hash, path);
+                    });
+                    this.blobsCachePendingHashes.add(hashP);
+                    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                    hashP.finally(() => {
+                        this.blobsCachePendingHashes.delete(hashP);
+                    });
+                }
+            }
+        }
+
+        this.ops = ops;
+        return id ? [{ id, treeId: undefined! }] : [];
+    }
+
+    private async fetchSnapshotWithMultiPartRequestCore(snapshotOptions: ISnapshotOptions, tokenFetchOptions: TokenFetchOptions) {
         const storageToken = await this.getStorageToken(tokenFetchOptions, "TreesLatest");
 
         // TODO: This snapshot will return deltas, which we currently aren't using. We need to enable this flag to go down the "optimized"
```

Comment:
is this TODO still relevant?

Reply:
Yes, this is not relevant now. Removed it.